### PR TITLE
CPPSDK : Add Async communication support and CPP code generation for x-uses + x-allow-focus=true

### PR DESCRIPTION
**Pull request by HaseenaSainul**

### CODE REVIEW

File: src/shared/modules.mjs [hunk 1]
```diff
@@ -289,6 +289,16 @@ const isPolymorphicReducer = compose(
     getPath(['tags'])
 )
 
+const isXUsesMethod = compose(
```
Comment (jlacivita) [inline src/shared/modules.mjs:292]: We should rename this `isAllowFocusMethod`.

Almost every method is an x-uses method, and this name really confused me when I was reading the code that calls it.

File: src/shared/modules.mjs [hunk 2]
```diff
@@ -359,6 +369,8 @@ const getPublicEvents = compose(
 const hasPublicInterfaces = json => json.methods && json.methods.filter(m => m.tags && m.tags.find(t=>t['x-provides'])).length > 0
 const hasPublicAPIs = json => hasPublicInterfaces(json) || (json.methods && json.methods.filter( method => !method.tags.find(tag => tag.name === 'rpc-only')).length > 0)
 
+const hasXUsesMethods = json => json.methods && json.methods.filter(m => isXUsesMethod(m)).length > 0
```
Comment (jlacivita) [inline src/shared/modules.mjs:372]: same here.

File: src/macrofier/engine.mjs [hunk 3]
```diff
@@ -108,7 +108,7 @@ const getTemplate = (name, templates) => {
 }
 
 const getTemplateTypeForMethod = (method, type, templates) => {
-  const name = method.tags && method.tags.map(tag => tag.name.split(":").shift()).find(tag => Object.keys(templates).find(name => name.startsWith(`/${type}/${tag}.`))) || 'default'
+  const name = method.tags ? (isXUsesMethod(method) && Object.keys(templates).find(name => name.startsWith(`/${type}/x-uses.`))) ? 'x-uses' : (method.tags.map(tag => tag.name.split(":").shift()).find(tag => Object.keys(templates).find(name => name.startsWith(`/${type}/${tag}.`)))) || 'default' : 'default'
```
Comment (jlacivita) [inline src/macrofier/engine.mjs:111]: We should rename this template `allowsFocus`

We don't include the `x-` that's just an OpenRPC convention. and see my other comments about allows focus vs uses.

**Review by jlacivita — CHANGES_REQUESTED** [review]

File: src/shared/modules.mjs (hunk 1 above)
Comment (HaseenaSainul) [reply src/shared/modules.mjs:292]: Sure, will update

File: src/shared/modules.mjs (hunk 2 above)
Comment (HaseenaSainul) [reply src/shared/modules.mjs:372]: Will update

File: src/macrofier/engine.mjs (hunk 3 above)
Comment (HaseenaSainul) [reply src/macrofier/engine.mjs:111]: will update

File: src/shared/modules.mjs (hunk 1 above)
Comment (HaseenaSainul) [reply src/shared/modules.mjs]: done

File: src/shared/modules.mjs (hunk 2 above)
Comment (HaseenaSainul) [reply src/shared/modules.mjs]: done

File: src/macrofier/engine.mjs (hunk 3 above)
Comment (HaseenaSainul) [reply src/macrofier/engine.mjs]: done

**Pull request merged**